tests/test_custom_components.py

```diff
@@ -0,0 +1,173 @@
+        "sortable": True,
+        "editable": False,
+        "tooltipComponent": "myCustomTooltip"
+    }
+
+    df2 = df.copy()
+    df2 = df2.to_dict('records')
+    df2[0]['quantity'] = 30
+
+    table = dag.AgGrid(
+        id="portfolio-grid",
+        className="ag-theme-alpine-dark",
+        columnDefs=columnDefs,
+        rowData=df.to_dict("records"),
+        columnSize="sizeToFit",
+        defaultColDef=defaultColDef,
+    )
+
+    header = html.Div("My Portfolio", className="h2 p-2 text-white bg-primary text-center")
+
+    app.layout = dbc.Container(
+        [
+            header,
+            dbc.Row(dbc.Col(table, className="py-4")),
+            html.Div(id='cellValueChanged')
+        ],
+    )
+
+    @app.callback(Output('cellValueChanged', 'children'), Input('portfolio-grid','cellValueChanged'), prevent_initial_call=True)
+    def showChange(n):
+        return json.dumps(n)
+
+    dash_duo.start_server(app)
+
+    grid = utils.Grid(dash_duo, "portfolio-grid")
+
+    grid.wait_for_cell_text(0, 1, "Apple")
+
+    ### testing components
+    assert grid.get_cell_html(0,0).strip() == """<a href="https://finance.yahoo.com/quote/AAPL" target="AAPL">AAPL</a>"""
+    assert grid.get_cell_html(0,
+                              3).strip() == """<div style="width: 100%; height: 100%; padding: 5px; display: flex; justify-content: center; align-items: center;"><div style="background-color: red; border-radius: 15px;">Low</div></div>"""
+    assert grid.get_cell_html(0,
+                              4).strip() == """<div style="width: 100%; height: 100%; padding: 5px; display: flex; justify-content: center; align-items: center;"><input type="checkbox" style="cursor: pointer;"></div>"""
+    assert grid.get_cell_html(0,
+                              5).strip() == """<div style="width: 100%; height: 100%; padding: 5px; display: flex; justify-content: center; align-items: center;"><button class="btn btn-success">buy</button></div>"""
+    assert grid.get_cell_html(0,
+                              6).strip() == """<div style="width: 100%; height: 100%; padding: 5px; display: flex; justify-content: center; align-items: center;"><button class="btn btn-danger">sell</button></div>"""
+    assert grid.get_cell_html(0,
+                             7).strip() == """<div style="width: 100%; height: 100%; padding: 5px; display: flex; justify-content: center; align-items: center;"><select><option>buy</option><option>sell</option><option>hold</option></select></div>"""
+    grid.element_click_cell_button(0,6)
+    time.sleep(1)
+    assert grid.get_element_text("#cellValueChanged") == '{"rowIndex": 0, "nodeId": "0", "data": {"ticker": "AAPL", "company": "Apple", "price": 154.98500061035156, "volume": "Low", "binary": false, "buy": {"children": "buy", "className": "btn btn-success"}, "sell": {"children": "sell", "className": "btn btn-danger", "n_clicks": null}, "action": "sell"}, "oldValue": "buy", "newValue": "sell", "colId": "action"}'
```

Comment:
Let's keep the `grid` class constrained to just things inside the `AgGrid` element. This can be just `dash_duo.wait_for_text_to_equal("#cellValueChanged", '{"rowIndex": 0, ...}'`